### DISCUSSION TITLE
Обсуждение проблем в ZCAD

### DISCUSSION THREAD

**Discussion by gtfox** [zamtmn/zcad#137] [General]
Иногда не понятно является ли то, что мы видим проблемой или багом, достойным оформления репорта.
Поэтому предлагаю тут это обсуждать

Записал видосик того, что заметил
https://disk.yandex.ru/i/NQ8gJNLobdrhFQ
выложил на диск, т.к. сюда видео до 10мб можно, а на дзене нет доступа к видео по ссылке, на рутубе есть, но там публикация сутки, ютуп загибается, а на платформу лень было лезть :)

**Comment by gtfox** [comment]
Заметил сейчас, и на видео тоже - падение произошло во время автосохранения

**Comment by gtfox** [comment]
Еще есть ошибки на многих устройствах 
![Анимация](https://github.com/user-attachments/assets/10f7852b-6010-444f-ba53-5ed4ed636f9b)

**Comment by vlsx** [comment]
Это проблема в определении свойств блока когда данные из DFX не совпадают с данными из PAS. Вернее файлов PAS программа не видит. Нужно время разобраться.

**Comment by zamtmn** [comment]
по видео
1 - текст отображается в разных случаях по разному, потому что он находится в разных местах. просто в чертеже или в динамической части блока. Сделано для ускорения перетаскивания больших количеств примитивов, как временное решение - но тянется уже долго(( оно зарепорчено - https://github.com/zamtmn/zcad/issues/87

2 - выделение после вставки. Чтобы примитивы не пришлось таскать после вставки надо указывать как вазовую точку при копировании, так и базовую точку при вставке. те. контрол-шыфт-C, контрол-V в остальных случаях двигать примитывы придется. ИМХО. не выделение дублированых примитивов я сделал исходя из соображений чтоб при повторном дублировании продублировались исходные примитивы, а не вновь вставленные.

3 - падения. Всё как я люблю - у меня не падает)) Ночная сборка смобирается релизным компилятором, релизы и то что использую я - собирается транковым. Гдето засела разница в поведении компилятора релиз\транк - пока ниче сказать немогу, выясняю

гифка тлже следствие пункта 3, в транке этой проблемы нет

**Reply by gtfox** [reply]
Возможно видео получилось сумбурным, поэтому напишу конкретно, что я хотел сказать:

0. После вставки (Ctrl+V или Ctrl+D) курсор должен поменяться и появиться примитив рядом с курсором. (сейчас не появляется, пока не двинешь мышь)
1. Скрытие текста из динамической части блока. Оно должно работать одинаково при Ctrl+V и при Ctrl+D. (сейчас при при Ctrl+D скрывается графика, а не текст)
2. Я тут написал сначала, потом переосмыслил и удалил :)
3. Падения будем ловить по мере возможности :)

По пункту 0-1:
Я понимаю, что это придирки, но расскажу почему...
Я вставляю примитив, и вижу, что ничего не меняется. Я думаю, что я промахнулся мимо Ctrl+V или клавишу недожал, хочу нажать Ctrl+V еще раз, начинаю шевелиться и вижу изменение курсора и появление примитива. Каждый раз ты на секунду задумываешься, что что-то не в порядке... Понятно, что к этому можно привыкнуть, но можно же добавить "refresh" курсора в конце процедур Ctrl+V и Ctrl+D

На счет скрытия текста: само скрытие меня не напрягает, особенно если оно увеличивает производительность!
Вопрос в другом: при Ctrl+V - скрывается текст, при Ctrl+D - скрывается примитив. По-моему тут что-то не то.

Всё это вроде мелочи, но в каждом моменте ты видишь не то, что ожидаешь. Если это не сложно поправить, то почему бы и нет :)

По второму пункту:
Я как-то не думал, что мы можем вставить не один примитив, а кусок схемы, и потом его передвинуть. И если он будет выделен будет проще...

> Чтобы примитивы не пришлось таскать после вставки

Согласен. Либо с базовой точкой либо двигать то, что вставилось
> чтоб при повторном дублировании продублировались исходные примитивы

согласен

**Reply by zamtmn** [reply]
Ctrl-D, Ctrl+V починил, но не до конца... тут пока не всё показывается, но курсор меняется)) Сейчас пропажа примитивов обусловлена недоделками в движке, пока думаю как проще будет выкрутится

**Comment by gtfox** [comment]
Тестирование ZCADElectrotech v0.9.14.0-7-g6c12db4a8
Автосохранение больше не падает, катушку реле можно двигать :), электроснабжение починилось
В целом всё отлично!

Заметил улёт обозначений для устройств из группы Схема
Если текст скрылся за экран - выноска пропадает
Выноска появляется - если перемещаться нажатием колеса мыши
![Анимация](https://github.com/user-attachments/assets/e7462b5d-f07c-46bf-a698-064f7d4bb8ab)


Падение при закрытии не сохранённого файла осталось
Для пробы закинул видосик в ВК (а то яндекс диск не резиновый :) )

https://vk.com/video29449643_456239113?list=ln-yEqCYmZ2pUZZBBUJ2Z

[zcadcrashreport.txt](https://github.com/user-attachments/files/17003853/zcadcrashreport.txt)
Закинул еще файл на котором проверял. Вдруг это он виноват.
[Без имени001.zip](https://github.com/user-attachments/files/17003854/001.zip)

Еще, по-моему, радианы остались в стилях текста
![01](https://github.com/user-attachments/assets/13127750-d044-4148-bf59-41d8bbfe7be1)

**Comment by zamtmn** [comment]
По выноскам я в курсе, но пока придется с этим мириться.
Вылет у меня воспроизвести не получилось. Без воспроизведения даже имея стек не всегда получается локализовать проблему

**Reply by gtfox** [reply]
> Вылет у меня воспроизвести не получилось.

Может с этим получится

https://github.com/user-attachments/assets/34bc84e4-95f5-48b7-9737-6d1769fa3250


[Без имени004.zip](https://github.com/user-attachments/files/17248981/004.zip)

**Comment by zamtmn** [comment]
>>Еще, по-моему, радианы остались в стилях текста

поправил, градусы работают. но система единиц исчисления чертежа не доделана, поэтому ничто кроме градусов работать не будет))

**Comment by gtfox** [comment]
В дополнение к этому вопросу https://github.com/zamtmn/zcad/discussions/137#discussioncomment-10835526
Заметил, это работает только на устройства из набора. Если копировать не повреждённые с чертежа - то нормально.
И если новый файл создать - то из набора нормально втавляются

https://github.com/user-attachments/assets/d413dc8c-edb4-4a5d-9b87-20f0c40d74eb

**Reply by zamtmn** [reply]
Определения блоков хранятся в файле, если ты сохранил файл в сбойной версии - возможно сломались определения блоков.

В 2х словах сейчас это работает так: зкад при старте сканирует папку preload и выполняет найденные там *.cmd файлы в которых прописаны команды загрузки "библиотечных" файлов с определениями блоков. зкад создает "невидимый" чертеж в который один за одним загружает "библиотечные" файлы.

Когда ты тыкаешь по катушке в дереве справа фактически выполняется команда "insert(device_sch_coil)" которая вставляет в чертеж устройство device_sch_coil

insert работает так: она смотрит в текущем чертеже есть ли определение требуемого блока\устройства
1. если есть - создает вставку на основе определения из текущего чертежа
2. если нет, смотрит есть ли определение требуемого блока в "невидимомм" "библиотечном" чертеже, если есть то копирует его в текущий чертеж и идет в п.1
3. если определения нет нигде, просто ругается про немыслимость происходящего

поэтому фактически вставка "библиотечных" устройств происходит только в самый первый раз, когда их определений еще нет на чертеже

в приложенном файле уже имеются сбойные определения устройств, они и вставляются

**Reply by gtfox** [reply]
Спасибо за подробный ответ, было полезно!
>ругается про немыслимость происходящего

ей бы корвалола попить )))

**Comment by gtfox** [comment]
Упала ночная сборка 0.9.14.0-65 https://ci.appveyor.com/api/buildjobs/pj37j8s5cd16gaan/artifacts/ZCADElectrotech-f39d205e8f6d1a553bf7d2ebc3dfac23db09b5c4.zip
во время рисования паутины
Надо офрмлять репорт или так пойдёт?
![01](https://github.com/user-attachments/assets/f801d76d-ba5e-4b39-8462-14bbac68cd86)
[zcad_windows.log](https://github.com/user-attachments/files/17306878/zcad_windows.log)
[zcadcrashreport.txt](https://github.com/user-attachments/files/17306881/zcadcrashreport.txt)
